extensions/incubator/src/main/java/io/opentelemetry/extension/incubator/trace/ExtendedBaggageBuilder.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.extension.incubator.trace;
+
+import io.opentelemetry.api.baggage.Baggage;
+import io.opentelemetry.api.baggage.BaggageBuilder;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import java.util.Map;
+
+public class ExtendedBaggageBuilder {
+  private final BaggageBuilder delegate;
+
+  private ExtendedBaggageBuilder(BaggageBuilder delegate) {
+    this.delegate = delegate;
+  }
+
+  public static ExtendedBaggageBuilder create(BaggageBuilder builder) {
+    return new ExtendedBaggageBuilder(builder);
+  }
+
+  public static ExtendedBaggageBuilder current() {
+    return create(Baggage.current().toBuilder());
+  }
+
+  public ExtendedBaggageBuilder set(String key, String value) {
+    delegate.put(key, value);
+    return this;
+  }
+
+  public ExtendedBaggageBuilder setAll(Map<String, String> baggage) {
+    baggage.forEach(delegate::put);
+    return this;
+  }
+
+  /**
+   * Set baggage items inside the given {@link SpanCallable}.
+   *
+   * @param spanCallable the {@link SpanCallable} to call
+   * @param <E> the type of the exception
+   * @return the result of the {@link SpanCallable}
+   */
+  public <T, E extends Throwable> T call(SpanCallable<T, E> spanCallable) throws E {
```

Comment:
I don't think this is right. This is a builder, and I think we should stick to builders being comprised of strictly setters and `build` method. Another problem with this is that it accepts a `SpanCallable`, but has nothing to do with spans. Maybe this could be fixed by renaming SpanCallable to something like ThrowingCallable, but let's go back to my original suggestion:

```
String result = Baggage.fromMap(Map.of("key", "value"))
  .storeInContext(Context.current())
  .wrap(() -> "result")
  .call();
```

You gave the [feedback](https://github.com/open-telemetry/opentelemetry-java/pull/6063#issuecomment-1846685251):

> it's not clear if my current baggage items are added or replaced
I have to understand that baggage is stored in Context - which I actually don't have to be aware of
I have to know that wrap comes before call - it's not necessarily long, but harder to discover

I think we should extend BaggageBuilder to include a `putAll` method (as mentioned above). This would remove the ambiguity about when items are added or replaced. 

To add, call `Baggage.current().toBuilder().putAll(Map.of("key", "value"))`
To replace (i.e. start with a clean slate), call `Baggage.empty().toBuilder().putAll(Map.of("key", "value"))`

I also think that we could extend Context with `wrapAndRun` and `wrapAndCall` methods (in addition to the existing `wrap` and `call` methods. This simplifies the call to something like:

```
String result = Baggage.current().toBuilder().putAll(Map.of("key", "value"))
  .storeInContext(Context.current())
  .wrapAndCall(() -> "result")
```

The user still has to know that baggage goes in context, but I actually think its better the user is forced to understand that up front. 

So taken together, my recommendations are:
- Add `putAll(Map<String, String>)` to `BaggageBuilder`
- Add `wrapAndRun(Runnable)`, `T wrapAndCall(Callable<T>)` to `Context`

If we add these, I think we can improve the ergonomics without needed to incubate anything at all.

Reply:
@jack-berg it was good that we split this out - this is more difficult to get right than I thought :smile: 

> its better the user is forced to understand that up front.

Why is that better? In this simple use case, it can be transparent to the user.

However, I'm fine with teaching the user - if we can find a way that doesn't hurt autocompletion/explorability.

In particular, this is what I guess users will stumble across:

- what is the "current" baggage vs. "empty"
- what is a "context"
- what is the "current" context
- what is "wrapping"

Reply:
How do you propose we proceed?